### PR TITLE
Refactor and fix WYSIWYG link issues

### LINKED ISSUE #15662: Unsaved Notebook insert link shows full path when editing existing relative link
1. Add text cell to Untitled Notebook
2. Create link `[test](./hello.txt)`
3. Select the link text in WYSIWYG mode and click the `Insert Link` button
4. Callout is opened but the path populated isn't relative - it seems to be from the root of the app 

e.g. `/C:/Users/chgagnon/AppData/Local/Programs/Azure%20Data%20Studio%20-%20RC1/hello.txt`

### LINKED ISSUE #15664: Fully qualified links inserted in WYSIWYG mode are turned into relative links
1. Create Notebook
2. Add text cell with link to local file
3. The link is turned into a relative path

e.g. 

`C:\Users\chgagnon\Desktop\test.txt` turns into `..\..\..\..\Desktop\test.txt`

This does not happen in split/MD mode.

### PR DESCRIPTION
**Pull request by barbaravaldez**
This PR fixes #15664 and fixes #15662. 
I did some refactoring to simplify the logic when we create html node and when we convert links to markdown.

### CODE REVIEW

File: src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	}
+}
+
+export function findPathRelativeToContent(notebookFolder: string, contentPath: URI | undefined): string {
+	if (notebookFolder) {
+		if (contentPath?.scheme === 'file') {
+			let relativePath = contentPath.fragment ? path.relative(notebookFolder, contentPath.fsPath).concat('#', contentPath.fragment) : path.relative(notebookFolder, contentPath.fsPath);
+			//if path contains whitespaces then it's not identified as a link
+			relativePath = relativePath.replace(/\s/g, '%20');
+			if (relativePath.startsWith(path.join('..', path.sep) || path.join('.', path.sep))) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts:82]: This check doesn't look correct, it'll evaluate `path.join('..', path.sep) || path.join('.', path.sep)` first but that's always going to just resolve to `path.join('..', path.sep)` (see || operator for strings). I assume you instead want to do 

`relativePath.startsWith(path.join('..', path.sep)) || relativePath.startsWith(path.join('.', path.sep))`

File: src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			case 'string': {
+				if (this._isFile && !this._isAbsolutePath && !this._isAnchorLink) {
+					const relativePath = (this._path).replace(/\\/g, path.posix.sep);
+					const linkUrl = path.resolve(this._notebookDirectory, relativePath);
+					return linkUrl;
+				}
+				return this._path;
+			}
+			case 'object': {
+				if (this._href) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts:37]: Why all these truthy checks? Are there values you expect to be empty/undefined sometimes?

File: src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import * as path from 'vs/base/common/path';
+
+export class MarkdownWYSIWYGLinkHandler {
+	private _isAnchorLink: boolean | undefined;
+	private _isFile: boolean | undefined;
+	private _isAbsolutePath: boolean | undefined;
+	private _notebookDirectory: string;
+	private _href: URI | undefined;
+
+	constructor(private _notebookURI: URI, private _path: string | HTMLAnchorElement) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts:11]: path doesn't really seem correct - maybe just `pathOrElement` just to be clearer?

File: src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
+import { URI } from 'vs/base/common/uri';
+import * as path from 'vs/base/common/path';
+
+export class MarkdownWYSIWYGLinkHandler {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts:4]: This class could use comments explaining what the different parts are supposed to be doing and why. For example `getLinkUrl` - what is this URL supposed to be for? What format is it going to be in? Is it going to be escaped or not? Stuff like that - just help make it so people know what to expect and don't have to read through the code for all this.

Comment (Charles-Gagnon) [reply src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts:4]: Is there anything specifically about this that makes it a "WYSIWYG Link Handler` vs just a generic link helper? Maybe just something like `NotebookLinkHelper`?

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]
These changes should definitely have tests - especially for the scenarios that you're fixing. (also you currently have one test that's broken)

File: src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	private _href: string | undefined;
+
+	constructor(private _notebookURI: URI, private _pathOrElement: string | HTMLAnchorElement) {
+		if (typeof this._pathOrElement === 'string') {
+			this._notebookLink = URI.parse(this._pathOrElement);
+			this._isFile = this._notebookLink.scheme === 'file';
+			this._isAbsolutePath = path.isAbsolute(this._pathOrElement);
+			this._isAnchorLink = this._pathOrElement.includes('#') && this._isFile;
+		} else {
+			this._notebookLink = this._pathOrElement.attributes['href']?.nodeValue ? URI.parse(this._pathOrElement.attributes['href']?.nodeValue) : undefined;
```
Comment (VasuBhog) [inline src/sql/workbench/contrib/notebook/browser/markdownWYSIWYGLinkHander.ts:19]: I was testing this branch against #15942 issue.
It seems that the absolute paths are working however it does not show the full absolute path (it should also include the scheme/drive it is stored in) ~ **C:\\Users** (unless this is intended)
image

Also, this does not seem to work properly for other link schemes HTTPS for example _www.google.com_
image
After inserting in WYSIWYG and changing to split view the link is not kept, and therefore does not work.

**Comment by VasuBhog** [comment]
I was testing this branch against #15942 issue. This seems to also fix the issue as we now respect absolute paths and do no convert them to a relative.

**Pull request closed without merging**